packages/mobile/src/identity/commentEncryption.ts

```diff
@@ -0,0 +1,309 @@
+): DecryptedComment {
+  Logger.debug(TAG + 'decryptComment', 'Decrypting comment')
+
+  if (!features.USE_COMMENT_ENCRYPTION || !comment || !commentKeyPrivate) {
+    Logger.debug(TAG + 'decryptComment', 'Invalid params, skipping decryption')
+    return { comment }
+  }
+
+  const { comment: decryptedComment, success } = decryptCommentRaw(
+    comment,
+    hexToBuffer(commentKeyPrivate),
+    isSender
+  )
+
+  if (success) {
+    Logger.debug(TAG + 'decryptComment', 'Comment decryption succeeded')
+    return extractPhoneNumberMetadata(decryptedComment)
+  } else {
+    Logger.error(TAG + 'decryptComment', 'Decrypting comment failed, returning raw comment')
+    return { comment }
+  }
+}
+
+export function embedPhoneNumberMetadata(
+  comment: string,
+  phoneNumberDetails?: PhoneNumberHashDetails
+) {
+  return phoneNumberDetails
+    ? comment + METADATA_CONTENT_SEPARATOR + phoneNumberDetails.e164Number + phoneNumberDetails.salt
+    : comment
+}
+
+export function extractPhoneNumberMetadata(commentData: string) {
+  const phoneNumMetadata = commentData.match(PHONE_METADATA_REGEX)
+  if (!phoneNumMetadata || phoneNumMetadata.length < 4) {
+    return { comment: commentData }
+  }
+
+  return {
+    comment: phoneNumMetadata[1],
+    e164Number: phoneNumMetadata[2],
+    salt: phoneNumMetadata[3],
+  }
+}
+
+interface IdentityMetadataInTx {
+  address: string
+  e164Number: string
+  salt: string
+  phoneHash?: string
+}
+
+// Check tx comments (if they exist) for identity metadata like phone numbers and salts
+export function* checkTxsForIdentityMetadata({ transactions }: NewTransactionsInFeedAction) {
+  try {
+    if (!transactions || !transactions.length) {
+      return
+    }
+    Logger.debug(TAG + 'checkTxsForIdentityMetadata', `Checking ${transactions.length} txs`)
+
+    const commentKeyPrivate: string | null = yield select(privateCommentKeySelector)
+    if (!commentKeyPrivate) {
+      Logger.error(TAG + 'checkTxsForIdentityMetadata', 'Missing comment key, should never happen.')
+      return
+    }
+
+    const newIdentityData = findIdentityMetadataInComments(transactions, commentKeyPrivate)
+
+    const verifiedMetadata: IdentityMetadataInTx[] = yield call(
+      verifyIdentityMetadata,
+      newIdentityData
+    )
+    yield call(updatePhoneNumberMappings, verifiedMetadata)
+    Logger.debug(TAG + 'checkTxsForIdentityMetadata', 'Done checking txs')
+  } catch (error) {
+    Logger.error(TAG + 'checkTxsForIdentityMetadata', 'Error checking metadata', error)
+    // Allowing error to be swallowed for now. Impact is just that tx may not be labelled with full contact info
+  }
+}
+
+// Check all transaction comments for metadata
+function findIdentityMetadataInComments(
+  transactions: TransactionFeedFragment[],
+  commentKeyPrivate: string
+) {
+  const newIdentityData: IdentityMetadataInTx[] = []
+  for (const tx of transactions) {
+    if (tx.__typename !== 'TokenTransfer' || tx.type !== TokenTransactionType.Received) {
```

Comment:
Do we want to skip `TokenTransactionType.Sent`?

Reply:
We do, no sense decrypting and processing your own identity data as you send it out :)

Reply:
Haha good point!